src/main/java/com/github/karmazyn/locking/LockerImpl.java

```diff
@@ -0,0 +1,65 @@
+package com.github.karmazyn.locking;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class LockerImpl implements Locker {
+
+    private final ConcurrentHashMap<Object, MutexImpl> locks = new ConcurrentHashMap<>();
+
+    @Override
+    public Mutex lock(Object object) {
+
+        if (object == null) {
+            throw new IllegalArgumentException();
+        }
+
+        boolean locked = false;
+        MutexImpl mutex = null;
+
+        while (!locked) {
```

Comment:
this line - 20, 44 and 55th are deeply connected, do you think there is better way to code this?

Reply:
If you moved locking into Mutex, and keep Locker as a lookup, the two classes would be more independent.